design/baremetal-operator/bmc-events.md

```diff
@@ -0,0 +1,151 @@
+faults or environmental changes quicker than they would get an alert
+through other channels. For example, some workloads may have a sidecar
+container that knows how to deal with an alert that a particular network
+interface went down, or that the CPU temperature reached a certain
+threshold.
+
+### Goals
+
+- Provide an API to manage subscriptions to events
+
+### Non-Goals
+
+- [Configurable events and thresholds](#configurable-events-and-thresholds)
+- Any kind of event polling
+- Software for processing the events, i.e. any webhook
+- BMC's beyond Redfish for now
+
+## Proposal
+
+### User Stories
+
+- I'd like to configure a my BMC to send events to a target URL.
+- I'd like to filter the types of events the BMC sends to my target URL.
+- I'd like to provide context to a particular event subscription.
+- I'd like to provide arbitrary headers.
+- I'd like the baremetal-operator to reconcile on the
+  BMCEventSubscription resource, and ensure it's state is accurate in
+  Ironic.
+
+## Design Details
+
+### Implementation Details
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: BMCEventSubscription
+metadata:
+  name: worker-1-events
+spec:
+   hostRef: ostest-worker-1
+   targetURI: https://events.apps.corp.example.com/webhook
+   filters:
+     - StatusChange
+     - ResourceAdded
+     - Alert
+   headerRef: webhookBridgeAuth
+   context: “SomeUserContext”
+status:
+  errorMessage: ""
+  errorCount: 0
+  subscriptionID: aa618a32-9335-42bc-a04b-20ddeed13ade
+```
+
+- A BMCEventSubscription resource represents a subscription to the events generated
+  by a specific BMC.
+- Ironic will manage configuring the subscription, using a new API for managing them.
+- The BMCEventSubscription with maintain a reference to a BareMetalHost.
+- The BMCEventSubscription will maintain a reference to the ironic
+  subscription ID.
+- The BMCEventSubscription will allow injection of headers using a
+  headerRef to a secret, for example to provide basic auth
+  credentials.
+- The baremetal-operator binary will be expanded to include 2
+  reconcilers, with dedicated controller/reconcile loops for
+  BareMetalHost and BMCEventSubscriptions.
+
+### Open Questions
+
+### Risks and Mitigations
+
+#### Thundering herd
+
+Large numbers of events across large numbers of BareMetalHosts could
+generate a lot of traffic, we provide users mitigation facilities by
+allowing them to filter events to specific types. Users can also control
+how much events their webhook receives by configuring the alert
+thresholds out of band.
+
+### Dependencies
+
+[Ironic Eventing API](https://storyboard.openstack.org/#!/story/2008366)
+needs to be complete.
+
+### Test Plan
+
+There are some existing POC code for working with Redfish Events, we
+could build on this to implement a test framework for BMC events. We could
+also consider modifying sushy-tools to support emulated eventing.
+
+### Upgrade / Downgrade Strategy
+
+Not required, this is a new API being introduced
```

Comment:
During an upgrade with this feature in use we will lose the Ironic database. How will we address that? We don't want to have duplicate registrations of web hooks on BMCs, for example.

Reply:
Adding a field in the "Status" specifying if the event registration at the BMC was successful might solve this issue?

Reply:
The real state is outside of the cluster (in the BMC), though, so if something else modifies the BMC configuration the status field would no longer reflect the real configuration.

Reply:
That's a good point, @dtantsur @iurygregory How does the Ironic side handle this if we adopt a host? Will you be able to import the subscriptions into Ironic so we could check the API for existing subscriptions?

Reply:
So, if I do remember in the ironic API we will offer create/delete/list subscriptions, I will try to check if the BMC's would accept duplicate registrations, Ironic can also check if the node has a subscription that matches the one you are trying to add.

Reply:
In my experimentation I found that the BMC would not report all of the details of subscriptions that were registered. I presume that was to protect values that might need to be kept private/secret. I had to encode a signature of the subscription in the description field so I could find the match again and avoid re-registering it. See https://github.com/dhellmann/redfish-event-controller/blob/main/controllers/redfish/eventsubscription_controller.go#L148 for details